bindings/src/main/java/io/pravega/storage/filesystem/FileSystemChunkStorage.java

```diff
@@ -0,0 +1,317 @@
+        Timer timer = new Timer();
+
+        if (handle.isReadOnly()) {
+            throw new IllegalArgumentException("Write called on a readonly handle of chunk " + handle.getChunkName());
+        }
+
+        Path path = Paths.get(config.getRoot(), handle.getChunkName());
+
+        long totalBytesWritten = 0;
+        try (FileChannel channel = getFileChannel(path, StandardOpenOption.WRITE)) {
+            long fileSize = channel.size();
+            if (fileSize != offset) {
+                throw new IndexOutOfBoundsException(String.format("fileSize (%d) did not match offset (%d) for chunk %s", fileSize, offset, handle.getChunkName()));
+            }
+
+            // Wrap the input data into a ReadableByteChannel, but do not close it. Doing so will result in closing
+            // the underlying InputStream, which is not desirable if it is to be reused.
+            ReadableByteChannel sourceChannel = Channels.newChannel(data);
+            while (length != 0) {
+                long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                assert bytesWritten > 0 : "Unable to make any progress transferring data.";
+                offset += bytesWritten;
+                totalBytesWritten += bytesWritten;
+                length -= bytesWritten;
+            }
+            channel.force(false);
+        } catch (IOException e) {
+            throwExeption(handle.getChunkName(), "doWrite", e);
+        }
+        return (int) totalBytesWritten;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException, UnsupportedOperationException {
+        try {
+            int totalBytesConcated = 0;
+            Path targetPath = Paths.get(config.getRoot(), chunks[0].getName());
+            long offset = chunks[0].getLength();
+
+            for (int i = 1; i < chunks.length; i++) {
+                val source = chunks[i];
+                Preconditions.checkArgument(!chunks[0].getName().equals(source.getName()), "target and source can not be same.");
+                Path sourcePath = Paths.get(config.getRoot(), source.getName());
+                long length = chunks[i].getLength();
+                Preconditions.checkState(offset <= getFileSize(targetPath));
+                Preconditions.checkState(length <= getFileSize(sourcePath));
+                try (FileChannel targetChannel = getFileChannel(targetPath, StandardOpenOption.WRITE);
+                     RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+                    while (length > 0) {
+                        long bytesTransferred = targetChannel.transferFrom(sourceFile.getChannel(), offset, length);
+                        offset += bytesTransferred;
+                        length -= bytesTransferred;
+                    }
+                    targetChannel.force(false);
```

Comment:
The `false` parameter here indicates that the file metadata change is not flushed. Given that we are changing the size of the file, I'd think that the metadata flush is necessary. What do you think?

Reply:
According to the fsync documentation, if file metadata is changed as a result of file data change then metadata is also forced sync. (any metadata change that affects ability to read data correctly). It will not however sync if the only change is related to non data attributes.
In this case, this is effectively same as force(true).  I'll make it true to avoid confusion.